keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+* Out of the box support by many logging backens (Elasticsearch, Stackdriver, BigQuery Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```json
+{"ts": "00:15:15.525108", "caller": "controller_utils.go:116", "msg": "Updating ready status", "core/node": {"name": "node-1"}, "core/pod": {"name": "nginx-1", "namespace": "default"}, "controller": {"status": false}}
+```
+
+#### Development format
+
+Dedicated for eyeballing by kubernetes developers. This format is expected to change with time as there is no need to maintain any guarantee for its compatibility. Currently proposed format is based on [glogr](https://github.com/go-logr/glogr). 
+
+Example:
+
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] controller “msg”=”Updating ready status” "core/node": {"name": "node-1"} "core/pod": {"name": "nginx-1", "namespace": "default"} "controller": {"status": false}
+```
+
+### Logging configuration
+
+ 
+A new flag will be introduced to all kubernetes components allowing users to select their preferred logging format and maintain backward compatibility.
+Proposed flag `--logging-format`, with 3 values:
+* `legacy` for old logging format
+* `json` for json production format
+* `dev` for development format
+
+Setting this flag will select particular logger implementation from registry. Any additional configuration needed by loggers should be provided by flags specific to its implementation. Default loggers should provide a basic customization, but it will not be encouraged to avoid feature bloat. Adding new loggers, changing logger behavior or extending its functionality should be done outside of Kubernetes by implementing logr API and registering it in code.
+
+We acknowledge that extension via code is not the best, but should be sufficient for now. 
+
+All logs from the `json` and `dev` loggers should go to standard output.  Out of the existing klog command line arguments `json` and `dev` logger should support ‘-v’ flag.  
+
+### Kubernetes log schema
+
+Introducing structured logging API is not enough to address one of the biggest issues with log creation or analysis, that is inconsistent metadata. Size of Kubernetes codebase could easily result in drifting naming conventions. As part of this proposal we want to propose a mechanism for ensuring consistency of metadata. To achieve this we will introduce a way to define a schema for subset of commonly used fields in kubernetes. This schema should improve reusability of fields and provide discovery mechanism for logging backend vendors.
+
+Introducing this schema will include three things: 
+* Schema definition format
+* Set of helper functions to populate this schema (e.g. `WithObject` for referencing k8s object)
+* Validation mechanism ensuring schema is not bypassed by lower level logging API
+
+TODO(serathius): Propose how definition format should look like, probably yaml similar to https://github.com/elastic/ecs/tree/master/schemas
+
```

Comment:
What I'm missing in this proposal is some discussion of what will happen to logs coming from k8s dependencies.
I think a bunch of those are actually using glog/klog currently. So we were at least compatible with them so far.

If we migrate to a different library now (with different configuration options), we will end up with two completely independent log stream, almost impossible to join one with the other.

Maybe I'm missing something important here and that's not a problem, but I think this is "must-have" piece for this KEP.
@serathius @44past4 @thockin

Reply:
klog v2 has an option to send all logs to provided logr instance. We want to use it for external dependencies. This is covered in last point of Preparation section and first point of Automatic migration section of Stage 1 / Alpha migration.